### PR TITLE
Automate creation of FedRAMP inventory (#3634)

### LINKED ISSUE #3634: Automate creation of FedRAMP inventory
From [Automating creation of a FedRAMP Integrated Inventory Workbook](https://aws.amazon.com/blogs/publicsector/automating-creation-fedramp-integrated-inventory-workbook/):

> Did you know AWS can help deliver an automated solution for creating the FedRAMP Integrated Inventory Workbook? This workbook needs to be updated and submitted to the FedRAMP Project Management Office (PMO) monthly for continuous monitoring. Automating this workbook saves manual work hours. Any customer going through the FedRAMP authorization process can leverage this workbook. Understand how to gather an inventory of AWS resources from AWS Config data to create the FedRAMP Integrated Inventory Workbook.

Edit 1: Include the list of packages installed in the GitLab instance.

Edit 2: Include accounts in the inventory so we can cover AC-j (account review).

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #3634 

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Deployed `dev.gitlab`
- [x] Deployed `anvildev.gitlab`
- [x] Modified schedules according to upgrade instructions on `dev.gitlab` 
- [x] Modified schedules according to upgrade instructions on `anvildev.gitlab`
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added checklist items to `prod` promotion (see UPGRADING.rst)
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by nadove-ucsc** [comment]
Decided at PL to use terraform to create all-regions aggregator to avoid region specificity

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ stages:
   - teardown
   - manual
   - sell_unused_slots
+  - create_fedramp_report
```
Comment (hannes-ucsc) [inline .gitlab-ci.yml]:
```suggestion
  - fedramp_inventory
```

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -126,3 +127,13 @@ sell_unused_slots:
     - schedules
   script:
     - python3 scripts/sell_unused_slots.py
+
+create_fedramp_report:
+  extends: .base
+  only:
+    - schedules
+  script:
+    - python3 scripts/create_fedramp_report.py SSP-A13-FedRAMP-Integrated-Inventory-Workbook-Template.xlsx report.xlsx
```
Comment (hannes-ucsc) [inline .gitlab-ci.yml]:
```suggestion
    - python3 scripts/compliance/fedramp_inventory.py scripts/compliance/fedramp_inventory.template.xlsx fedramp_inventory.xlsx
```

File: requirements.txt [hunk 3]
```diff
@@ -15,6 +15,7 @@ google-api-core==2.10.2
 http-message-signatures==0.4.4
 jmespath==1.0.1
 more-itertools==9.0.0
+openpyxl==3.0.6
```
Comment (hannes-ucsc) [inline requirements.txt]: This should be a development dependency, not a runtime one.

File: src/azul/fedramp_report_service.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+                            unique_id=resource['configuration']['instanceId'],
+                        )
+
+    def _get_ip_address(self, ip_addresses: JSON, *keys) -> str:
+        for key in keys:
+            ip_addresses = ip_addresses[key]
+        return ip_addresses
+
+
+class ElbMapper(Mapper):
```
Comment (hannes-ucsc) [inline src/azul/fedramp_report_service.py]:
```suggestion
class ELBMapper(Mapper):
```

File: src/azul/fedramp_report_service.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    def _get_ip_addresses(self, availability_zones: JSONs) -> set[Optional[str]]:
+        return {
+            load_balancer_address.get('ipAddress')
+            for availability_zone in availability_zones
+            for load_balancer_addresses in availability_zone.get('loadBalancerAddresses', ())
+            for load_balancer_address in load_balancer_addresses
+        }
+
+
+class RdsMapper(Mapper):
```
Comment (hannes-ucsc) [inline src/azul/fedramp_report_service.py]:
```suggestion
class RDSMapper(Mapper):
```

and so on

File: src/azul/fedramp_report_service.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+            software_vendor='AWS',
+            system_owner=self._get_owner(resource),
+            unique_id=resource['arn'],
+        )
+
+
+class ElasticSearchMapper(Mapper):
+
+    def _supported_resource_types(self) -> set[str]:
+        return {'AWS::Elasticsearch::Domain'}
```
Comment (hannes-ucsc) [inline src/azul/fedramp_report_service.py]: Interesting, Amazon didn't rename this?

File: scripts/create_fedramp_report.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import argparse
+import logging
+import pathlib
+import sys
+
+from azul.fedramp_report_service import (
```
Comment (hannes-ucsc) [inline scripts/create_fedramp_report.py]:
```suggestion
from azul.compliance.fedramp_inventory_service import (
```

File: scripts/create_fedramp_report.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    FedRAMPReportService,
```
Comment (hannes-ucsc) [inline scripts/create_fedramp_report.py]:
```suggestion
    FedRAMPInventoryService,
```

File: .gitlab-ci.yml (hunk 2 above)
Comment (hannes-ucsc) [reply .gitlab-ci.yml]: The output file should probably be .gitignored.

 I think we should hard-code the template path in the script and only parameterize the output location.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/fedramp_report_service.py (hunk 6 above)
Comment (nadove-ucsc) [reply src/azul/fedramp_report_service.py]: https://docs.aws.amazon.com/config/latest/developerguide/resource-config-reference.html#amazonopensearchservice

Comment (nadove-ucsc) [reply src/azul/fedramp_report_service.py]: Listing the resources in `us-east-1` shows 2 instances of `AWS::Elasticsearch::Domain` and none of `AWS::OpenSearch::Domain`:

```
2023-03-06 15:55:19,813 DEBUG   MainThread: Mapping 'AWS::Elasticsearch::Domain' resource using 'ElasticSearchMapper'
2023-03-06 15:55:19,813 DEBUG   MainThread: Mapped to 1 rows
2023-03-06 15:55:19,813 DEBUG   MainThread: Mapping 'AWS::Elasticsearch::Domain' resource using 'ElasticSearchMapper'
2023-03-06 15:55:19,814 DEBUG   MainThread: Mapped to 1 rows
```

**Comment by nadove-ucsc** [comment]
Reviewers: please note that some aspects of the PR have deliberately been left incomplete per @hannes-ucsc 's instructions, as they need to be tested/developed after the changes are merged. The failing coverage check is also expected.

**Comment by theathorn** [comment]
Skipping peer review.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://coverage.readthedocs.io/en/6.4.4/excluding.html

This will probably fail on GitLab due to missing AWS permissions. Try to anticipate as best as you can what those would be and add them to the `aws_iam_policy_document.gitlab_iam` of the `gitlab` TF component.

**Comment by nadove-ucsc** [comment]
Filled in checklist for Peer Review.

**Pull request merged**